Refine Endpoint selection for Multi-cluster Service

luolanzone opened the pull request:
In order to decouple Multi-cluster with Antrea Proxy and refine Endpoint selection for Multi-cluster Service, following changes are included in this PR:

1. Revert previous changes for local Endpoint selection.
2. Add a new flow for the Service's ClusterIP in the EndpointDNAT table with
group action. When an Endpoint of a Multi-cluster Service is a local Service
ClusterIP and being selected, it will go to the corresponding exported Service's
group to select the final Endpoint. This can avoid that the traffic goes out of the
OVS bridge from antrea-gw0 (and handled by kube-proxy when it is running) and
comes back again.

The proposal details can be found in the comment:
https://github.com/antrea-io/antrea/pull/4508#discussion_r1130308109

luolanzone committed:
Revert "Refine Endpoint selection for multi-cluster Service (#4508)"

This reverts commit 6cdbca37e062a7b7ac9e466c2459b73caa7e0817.

Signed-off-by: Lan Luo <luola@vmware.com>

luolanzone commented:
@jianjuns @tnqn could you take a look at this PR, I am still working on unit test. thanks.
Btw, should I merge the reverted commit and new commit into one? or keep the reverted commit standalone?

tnqn commented on pkg/agent/openflow/client.go:
```diff
@@ -104,10 +103,14 @@ type Client interface {
 
+	// InstallServiceClusterIPFlows install flows for accessing Endpoints which is Service's ClusterIP, and has an action
+	// to Service's corresponding group.
+	InstallServiceClusterIPFlows(svcIP net.IP, svcPort uint16, protocol binding.Protocol, groupID binding.GroupIDType) error
```
Thought it over again, perhaps it's simpler to just generate one more flow in `InstallServiceFlows ` directly.

```
func InstallServiceFlows() {
...
if svcType == v1.ServiceTypeClusterIP && !nested {
      flows = append(flows, c.featureService. endpointRedirectFlowForServiceIP(svcIP, svcPort, protocol, groupID))
}
...
}

```

luolanzone replied:
Make sense, I will update it.

tnqn commented on pkg/agent/openflow/fields.go:
```diff
@@ -125,7 +125,9 @@ var (
-	TrafficControlActionField     = binding.NewRegField(4, 22, 23)
+	TrafficControlActionField = binding.NewRegField(4, 22, 23)
+	// reg4[24]: Mark to indicate whether the Service is a Multi-cluster Service.
+	MulticlusterServiceRegMark    = binding.NewOneBitRegMark(4, 24)
```
no need to be specific multicluster, perhaps just NestedServiceRegMark

tnqn commented on pkg/agent/proxy/proxier.go:
```diff
@@ -112,19 +108,18 @@ type proxier struct {
+	endpointSliceEnabled             bool
+	proxyLoadBalancerIPs             bool
+	topologyAwareHintsEnabled        bool
+	serviceWithNestedEndpointSupport bool
```
```suggestion
	supportNestedService bool
```

tnqn commented on pkg/agent/proxy/proxier.go:
```diff
@@ -157,6 +152,17 @@ func (p *proxier) removeStaleServices() {
+			if !exists {
+				continue
+			}
+			if err := p.ofClient.UninstallServiceClusterIPFlows(svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol, groupID); err != nil {
```
This is not needed if we merge the flow to InstallServiceFlows.

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -558,9 +534,15 @@ func (p *proxier) installServices() {
 			// Install ClusterIP flows for the Service.
 			groupID := p.groupCounter.AllocateIfNotExist(svcPortName, internalPolicyLocal)
-			if err := p.ofClient.InstallServiceFlows(groupID, svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol, uint16(affinityTimeout), externalPolicyLocal, corev1.ServiceTypeClusterIP); err != nil {
+			if err := p.ofClient.InstallServiceFlows(groupID, svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol, uint16(affinityTimeout), externalPolicyLocal, corev1.ServiceTypeClusterIP, hasNestedService); err != nil {
```
I think the new parameter should be a pointer

- nil, do nothing in `InstallServiceFlows`. The Service is not related to nested Service, like NodePort or LoadBalancer.
- true, install extra flow in EndpointDNATTable. The Service could be used as Endpoint of MC Service.
- false, add extra action NestedServiceRegMark to the flow in ServiceLBTable. The Service is a MC Service.

In addition, this pointer should be always nil when MC is not enabled.

hongliangl replied:
I remember that you said at least now  only ClusterIP can be used as the Endpoint of MC Service according to Kubernetes KEP-xxxx. If so, I think we should only enhance the flow of pure ClusterIP Service and skip the ClusterIP in NodePort or LoadBalancer.

luolanzone replied:
Yes, I mean `ClusterIP`, not ClusterIP type of Service, since both NodePort and LoadBalancer have ClusterIP, I mean ClusterIP can be exported instead of `NodeIP:NodePort` or "LBIP:Port". KEP-1645 also suggested to use ClusterIP of NodePort and LoadBalancer Services.

hongliangl replied:
Could we support that using ClusterIP of NodePort or LoadBalancer as Endpoint of MC-Service currently?

hongliangl commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -2410,9 +2407,33 @@ func (f *featureService) serviceLBFlow(groupID binding.GroupIDType,
 	if f.enableAntreaPolicy {
 		flowBuilder = flowBuilder.Action().LoadToRegField(ServiceGroupIDField, uint32(groupID))
 	}
+	if nested {
+		flowBuilder = flowBuilder.Action().LoadRegMark(NestedServiceRegMark)
+	}
```
Maybe add some comments here. This reg mark is loaded by MC Service, not the normal Service. I thought of it is loaded by normal Service since the normal Service is nested in MC Service.

luolanzone replied:
Done

tnqn reviewed (commented):
LGTM overall, there are integration test failures.

luolanzone commented on pkg/agent/openflow/client.go:
```diff
@@ -693,14 +693,17 @@ func (c *client) UninstallEndpointFlows(protocol binding.Protocol, endpoint prox
 	if affinityTimeout != 0 {
 		flows = append(flows, c.featureService.serviceLearnFlow(groupID, svcIP, svcPort, protocol, affinityTimeout, nodeLocalExternal, svcType))
 	}
+	if svcType == v1.ServiceTypeClusterIP && !nested {
+		flows = append(flows, c.featureService.endpointRedirectFlowForServiceIP(svcIP, svcPort, protocol, groupID))
+	}
```
Yes, the purpose is not to introduce too many Multi-cluster specific logic into Proxy. We may refine this part if there is any potential performance issue in the future.

luolanzone commented on pkg/agent/openflow/fields.go:
```diff
@@ -125,7 +125,9 @@ var (
 	// reg4[22..23]: Field to store the action of a traffic control rule. Marks in this field include:
-	TrafficControlActionField     = binding.NewRegField(4, 22, 23)
+	TrafficControlActionField = binding.NewRegField(4, 22, 23)
+	// reg4[24]: Mark to indicate whether the Service has Endpoints which is other Service's ClusterIP.
```
Changed to `Mark to indicate whether the Endpoints of a Service includes other Service's ClusterIP`

luolanzone commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -2410,9 +2407,33 @@ func (f *featureService) serviceLBFlow(groupID binding.GroupIDType,
 	return flowBuilder.Action().Group(groupID).Done()
 }
 
+func (f *featureService) endpointRedirectFlowForServiceIP(clusterIP net.IP, svcPort uint16, protocol binding.Protocol, groupID binding.GroupIDType) binding.Flow {
```
Done

luolanzone commented on pkg/agent/proxy/proxier.go:
```diff
@@ -558,9 +534,15 @@ func (p *proxier) installServices() {
+			// Check the `NestedServiceSupport` field only when Proxy is enabled with `supportNestedService`.
+			if p.supportNestedService {
+				hasNestedService = svcInfo.NestedServiceSupport
+			}
```
Done

luolanzone committed:
Refine Endpoint selection for MC Service

Add a new flow for the Service's ClusterIP in the EndpointDNAT table with
group action. When an Endpoint of a Multi-cluster Service is a local Service
ClusterIP and being selected, it will go to the corresponding exported Service's
group to select the final Endpoint. This can avoid that the traffic goes out of the
OVS bridge from antrea-gw0 (and handled by kube-proxy when it is running) and
comes back again.

The proposal details can be found in the comment:
https://github.com/antrea-io/antrea/pull/4508#discussion_r1130308109

Signed-off-by: Lan Luo <luola@vmware.com>

tnqn commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -2410,9 +2407,36 @@ func (f *featureService) serviceLBFlow(groupID binding.GroupIDType,
 		flowBuilder = flowBuilder.Action().LoadToRegField(ServiceGroupIDField, uint32(groupID))
 	}
+	if nested {
+		// `nested` is true only when a Service is a Multi-cluster Service for now.
```
I feel the comment belongs to where nested is calculated here, adding here still makes it specific to Multi-cluster while it doesn't have to.

luolanzone replied:
Done

tnqn replied:
don't see any change

tnqn commented on pkg/agent/proxy/types/types.go:
```diff
@@ -22,16 +22,24 @@ import (
 	*k8sproxy.BaseServiceInfo
 	// cache for performance
 	OFProtocol openflow.Protocol
+	// NestedServiceSupport means the Service has an annotation "multicluster.antrea.io/imported-service"
+	// which means its Endpoints might be another Serivce's ClusterIP.
+	NestedServiceSupport bool
```
```suggestion
	// IsNested means the Service's Endpoints could be another Service's ClusterIP.
	// Currently it's true for Multicluster Service, determined by whether there is a Multicluster specific annotation.
	IsNested bool
```

luolanzone replied:
Done

hongliangl commented on pkg/agent/openflow/client.go:
```diff
@@ -693,14 +693,17 @@ func (c *client) UninstallEndpointFlows(protocol binding.Protocol, endpoint prox
 	if affinityTimeout != 0 {
 		flows = append(flows, c.featureService.serviceLearnFlow(groupID, svcIP, svcPort, protocol, affinityTimeout, nodeLocalExternal, svcType))
 	}
+	if svcType == v1.ServiceTypeClusterIP && !nested {
+		flows = append(flows, c.featureService.endpointRedirectFlowForServiceIP(svcIP, svcPort, protocol, groupID))
+	}
```
In this way, I think we can skip the ClusterIP flow of NodePort or LoadBalancer.

luolanzone replied:
As I mentioned in previous comment, I feel there is no need to skip it. They will be be supported via ClusterIP.

luolanzone replied:
I think the only gap here is that we didn't officially announce that we support these two kinds of Service in multi-cluster docs,  I probably will update it after double check MC controller codes.

hongliangl replied:
If we don't officially announce and support that , maybe we shouldn't introduce extra flows which are not used totally.

luolanzone replied:
As I mentioned, there are probably just doc updates involved to support NodePort or LoadBalancer with ClusterIP, so I prefer not to skip it, otherwise we may change this part back soon. If you have further concern, we can sync offline, thanks.

hongliangl replied:
If ClusterIP from NodePort or LoadBalancer will be supported eventually, I'm ok with that.

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -558,9 +534,15 @@ func (p *proxier) installServices() {
 				}
 			}
 
+			var isNestedService bool
+			if p.supportNestedService {
+				// Check the `IsNested` field only when Proxy is enabled with `supportNestedService`.
+				isNestedService = svcInfo.NestedServiceSupport
+			}
```
```go
var isNestedService *bool
if p.supportNestedService && svcInfo.NodePort() == 0 {
	isNestedService = &svcInfo.NestedServiceSupport
}
```

luolanzone commented:
/test-multicluster-e2e

jianjuns commented on pkg/agent/openflow/fields.go:
```diff
@@ -125,7 +125,9 @@ var (
 	// reg4[22..23]: Field to store the action of a traffic control rule. Marks in this field include:
-	TrafficControlActionField     = binding.NewRegField(4, 22, 23)
+	TrafficControlActionField = binding.NewRegField(4, 22, 23)
+	// reg4[24]: Mark to indicate whether the Endpoints of a Service includes other Service's ClusterIP.
```
other -> another

luolanzone replied:
Done

luolanzone commented on pkg/agent/proxy/types/types.go:
```diff
@@ -22,16 +22,25 @@ import (
 	k8sproxy "antrea.io/antrea/third_party/proxy"
 )
 
+const AntreaMCServiceAnnotation = "multicluster.antrea.io/imported-service"
```
Done, add a new func.

luolanzone commented:
/test-multicluster-e2e

luolanzone commented:
/test-all

jianjuns commented on multicluster/controllers/multicluster/common/helper.go:
```diff
@@ -104,3 +104,7 @@ func HashLabelIdentity(l string) string {
 }
+
+func IsMulticlusterService(service *corev1.Service) bool {
+	return service.Annotations[AntreaMCServiceAnnotation] == "true"
```
Where `AntreaMCServiceAnnotation `is defined?

luolanzone replied:
I found it's already defined in `multicluster/controller/multicluster/common/types.go`, so just reuse it.

tnqn reviewed (dismissed):
LGTM

jianjuns commented:
@hongliangl : any further comments?
@luolanzone : please resolve the conflict.

luolanzone commented:
Conflicts resolved.

/test-multicluster-e2e
/test-all

tnqn commented:
/test-multicluster-e2e

jianjuns commented:
/test-multicluster-e2e

luolanzone commented:
Seems the testbed is not stable again, the e2es take much longer time which is 2 times than before, and NP tests failed with timeout, I will check with @hjiajing.

luolanzone commented:
/test-multicluster-e2e

luolanzone commented:
/test-multicluster-e2e

luolanzone commented:
/test-multicluster-e2e

luolanzone commented:
/test-multicluster-e2e

luolanzone commented:
/test-multicluster-e2e

luolanzone commented:
I have tried to trigger MC e2e locally with this branch, the results looks fine, all tests are passed. The latest triggered jenkins job might be back to normal. @tnqn you may either merged this or wait for the latest build result from Jenkins. Thanks.
```
--- PASS: TestConnectivity (340.06s)
    --- PASS: TestConnectivity/TestMCService (63.24s)
        --- PASS: TestConnectivity/TestMCService/Case=MCServiceConnectivity (0.75s)
        --- PASS: TestConnectivity/TestMCService/Case=ScaleDownMCServiceEndpoints (3.02s)
        --- PASS: TestConnectivity/TestMCService/Case=ANPToServices (12.61s)
        --- PASS: TestConnectivity/TestMCService/Case=StretchedNetworkPolicy (12.62s)
        --- PASS: TestConnectivity/TestMCService/Case=StretchedNetworkPolicyReject (6.66s)
        --- PASS: TestConnectivity/TestMCService/Case=StretchedNetworkPolicyUpdatePod (3.54s)
        --- PASS: TestConnectivity/TestMCService/Case=StretchedNetworkPolicyUpdateNS (6.67s)
        --- PASS: TestConnectivity/TestMCService/Case=StretchedNetworkPolicyUpdatePolicy (3.35s)
    --- PASS: TestConnectivity/TestAntreaPolicy (264.22s)
PASS
ok  	antrea.io/antrea/multicluster/test/e2e	342.245s
```
